File: ray-tracing-web/src/source/Vector.ts
export class Vector {
    public constructor(
        public readonly x: number,
        public readonly y: number,
        public readonly z: number,
    ) {
    }

    public times(k: number): Vector {
        return new Vector(k * this.x, k * this.y, k * this.z);
    }

    public minus(other: Vector): Vector {
        return new Vector(this.x - other.x, this.y - other.y, this.z - other.z);
    }

    public plus(other: Vector): Vector {
        return new Vector(this.x + other.x, this.y + other.y, this.z + other.z);
    }

    public dot(other: Vector): number {
        return this.x * other.x + this.y * other.y + this.z * other.z;
    }

    public mag2(): number {
        return this.x * this.x + this.y * this.y + this.z * this.z;
    }

    public mag(): number {
        return Math.sqrt(this.mag2());
    }

    public norm(): Vector {
        return this.times(1 / this.mag());
    }

    public cross(other: Vector): Vector {
        return new Vector(
            this.y * other.z - this.z * other.y,
            this.z * other.x - this.x * other.z,
            this.x * other.y - this.y * other.x,
        );
    }
}

File: ray-tracer/common/Vector.cs
using System;

namespace ray_tracer.common
{
    class Vector
    {
        public readonly double x;

        public readonly double y;

        public readonly double z;

        public Vector(double x, double y, double z)
        {
            this.x = x;
            this.y = y;
            this.z = z;
        }

        public Vector times(double k)
        {
            return new Vector(k * x, k * y, k * z);
        }

        public Vector minus(Vector other)
        {
            return new Vector(x - other.x, y - other.y, z - other.z);
        }

        public Vector plus(Vector other)
        {
            return new Vector(x + other.x, y + other.y, z + other.z);
        }

        public double dot(Vector other)
        {
            return x * other.x + y * other.y + z * other.z;
        }

        public double mag2()
        {
            return x * x + y * y + z * z;
        }

        public double mag()
        {
            return Math.Sqrt(mag2());
        }

        public Vector norm()
        {
            return times(1 / mag());
        }

        public Vector cross(Vector other)
        {
            return new Vector(
                y * other.z - z * other.y,
                z * other.x - x * other.z,
                x * other.y - y * other.x
            );
        }
    }
}
